### ISSUE TITLE
Help with: REST interface to fabric

### ISSUE THREAD

**Issue by erikgunther** [fabric/fabric#1034]
Hi all, 

This will probably be a bit longer story to read, but I do have some questions at the bottom, that I hoping to get some help with.
## Background

We have a setup tool for setting up our dev. environment its rather complex, currently only on localhost, but I'm guessing that we will use it for test environments on eg. AWS later. 

Anyway as I try to make the code reader friendly and user friendly it have many tasks and it do keep track of dependencies between tasks (by using execute(<other_task>)). There are many task to choose from and that do confuse the users of the tool a bit. The idea popped up to setup a web-interface showing all the tasks and let the user point and click and to show all dependencies to let them easier know what is happening, when they run task X.

So during one of the HackWeek we have at the company I decided to do this, with some help from a coworker.
## The Idea

Have a REST interface to fabric that will execute the tasks given from the web-interface and send the feedback back to the browser too. Also as Fabric can intercept password request, send these back to the web page and let the user handle them there.
## The Implementation

My idea here was to use Flask as the REST interface into fabric, connect to the inner beings in fabric and add a job-handler so the interface become something like this:

 /list -- list all task (cmd: fab -l)
 /task/<task> - Execute task, return id  (cmd: fab task)
 /status/<id> - Print and interact with the task (cmd: view and type when needed.)

All is responding with a json-documents.

This is all done in fabric/rest.py also added a fab-server currently with the same command line interface as fab.
## Web interface

This part is as I see it is always custom so its up to each user of fabric to develop this.But I will happily add some documentation about how we have done this. We are using handelbars.js 
## The Problem

Unfortunately I didn't really get all the way mostly due to lack of understanding of fabric.io. 

In my branch https://github.com/erikgunther/fabric/tree/feature/REST I only execute via process.Popen('fab <task>') and that isn't good enough, but it works for showing the idea. The question here is how to piggy back on fabric.io to redirect stdout, stderr and stdin to the webpage. I couldn't get my head around and my test failed for replacing the paramiko.channel object, that fabric.io is using. If you could help me understand how to redirect std\* will running  "execute('task')" I think I can pick up from there.

This implementation isn't complete nor ready for merge, I missing many of the state change parameters lite host, users and so on but I recon these will be quite easy to add when I can execute within the same fabric context. But as always each call to /task will generate a thread/subprocess with local environment.

/BR Erik

**Comment by cgarciaarano** [comment]
Hi,

I've been working in something similar, using fabric as library (as @hummermania suggested). It's not mature at all, I'm just testing stuff, but I have faced the same problems. Now I have a website that can send commands to the _fabric server_ via WebSockets, and catch the output, posting it back to the browser. It's not a REST interface, it's a website, but the foundations are there. This [tutorial](http://blog.miguelgrinberg.com/post/easy-websockets-with-flask-and-gevent) about Flask + WebSockets was inspiring.

![image](https://cloud.githubusercontent.com/assets/6098044/7705525/020508bc-fe45-11e4-830f-b3f0400953b0.png)


So, instead of using `Popen`, I'm using `execute`. `execute` allows to pass a `StringIO` as `stdout`. When the `execute` call has finished, you just need to get the value of the `StringIO` to get the output of the `execute` command.

In a nutshell:
```python
import cStringIO

io_buffer = cStringIO.StringIO()
execute(your_task, params, stdout = io_buffer, hosts = ['localhost'])

print io_buffer.getvalue()
```

Things become trickier when you want to print the stdout as it's being written, but I let you dive in it as my solution is pretty lame.

Now I'm facing the problem of `stdin`, as `execute` doesn't include `stdin` as a parameter, and I don't want to change `fabric` behavior (it's used as a library). If everything goes well and my project gets to version 0.1, I will make it public.

Regards

**Comment by kaaveland** [comment]
I looked around for similar projects to see if there were any, maybe https://github.com/kevin1024/fabric_remote has solved this problem in some novel way?

**Comment by kaaveland** [comment]
I'm thinking that this sort of question would get more attention on the mailing list, it might be worth posting your question there: http://www.fabfile.org/contact.html

**Comment by bitprophet** [comment]
Gonna leave this open if only because it comes up pretty often. I know I've seen multiple solutions (specific to Fabric and generic, such as "just use Jenkins" [which is what my current employer does]) but never took the time to note them down. We should probably add a doc page that lists such things, at which point we could close this.
